Implement local signing of ethereum transactions inside the KM enclave.

reuvenpo opened the pull request:
This PR by @Cashmaney builds on @reuvenpo's PR #247 which adds local signing of keys stored in the KM configuration JSON file, and adds an option to sign all transactions inside the KM enclave, using keys generated and stored there.

reuvenpo commented on .dockerignore:
```diff
@@ -0,0 +1,12 @@
+.idea
+.github
+dockerfile
+Cargo.lock
```
Consider if we should remove this and the `Dockerfile` files

reuvenpo commented on enigma-core/app/Cargo.toml:
```diff
@@ -34,7 +33,7 @@ sgx_urts = { git = "https://github.com/baidu/rust-sgx-sdk.git", rev = "v1.0.9" }
 [dev-dependencies]
 ethabi = "8.0.1"
-cross-test-utils = { path = "cross-test-utils" }
+# cross-test-utils = { path = "cross-test-utils" }
```
Please remove this line if it isn't used.

reuvenpo reviewed (commented):
Please revert all changes made to files in `enigma-core/app` and `enigma-core/enclave`

reuvenpo commented on enigma-crypto/src/asymmetric.rs:
```diff
@@ -124,6 +124,20 @@ impl KeyPair {
     }
 
+    /// Same as the function above, but without hashing, and we have to get a 32 byte message
+    pub fn sign_hashed(&self, message: &[u8; 32]) -> Result<[u8; 65], CryptoError> {
```
Please define the `sign` method in terms of this method

reuvenpo commented on enigma-crypto/src/lib.rs:
```diff
@@ -58,3 +58,8 @@ pub trait Encryption<T, E, R, N>
     fn decrypt(enc: R, key: T) -> Result<Self, E>;
 }
 
+/// This object knows how to sign like a motherfucker
```
please provide a more descriptive description

reuvenpo commented on enigma-crypto/src/lib.rs:
```diff
@@ -58,3 +58,8 @@ pub trait Encryption<T, E, R, N>
 
+/// This object knows how to sign like a motherfucker
+pub trait EcdsaSign {
+    /// This object knows how to sign like a motherfucker
```
please

reuvenpo commented on enigma-principal/app/src/esgx/equote.rs:
```diff
@@ -29,6 +35,28 @@ pub fn get_register_signing_address(eid: sgx_enclave_id_t) -> Result<[u8; 20], E
+    }
+}
+
+// wrapper function for creating a signature using the ethereum key
```
use a triple `///` to mark function descriptions as formal documentation.

reuvenpo commented on enigma-principal/app/src/cli/app.rs:
```diff
@@ -1,32 +1,67 @@
+
+        path.pop();
+        path.push("ethereum-account-addr.txt");
+        let mut file = File::create(path.clone())?;
```
spare the clone, you can just `File::create(&path)?`. apply this pattern to all the clones of this object in this scope.

reuvenpo commented on enigma-principal/app/src/cli/app.rs:
```diff
@@ -1,32 +1,67 @@
         println!("Wrote signing address: {:?} in file: {:?}", prefixed_signing_address, path);
+
+        path.pop();
+        path.push("ethereum-account-addr.txt");
```
What do you do with this file?

reuvenpo commented on enigma-principal/app/src/cli/app.rs:
```diff
@@ -1,32 +1,67 @@
     let mut principal_config = PrincipalConfig::load_config(opt.principal_config.as_str())?;
     let report_manager = ReportManager::new(principal_config.clone(), eid)?;
     let signing_address = report_manager.get_signing_address()?;
+    let ethereum_address = report_manager.get_ethereum_address()?;
```
In this scope, it's unclear what the difference is between `signing_address` and `ethereum_address`.

reuvenpo commented on enigma-principal/app/src/boot_network/principal_manager.rs:
```diff
@@ -76,6 +116,15 @@ impl ReportManager {
+        let _signing_address = esgx::equote::get_ethereum_address(self.eid)?;
+        let signing_address = _signing_address.to_vec().to_hex();
+        Ok(signing_address)
+    }
```
Arguably, this method should return the raw ethereum address as a [u8; 20] (or, preferably, an actual type that represents it, probably from one of our dependencies). It can later be cast to its hex representation when needed.

Cashmaney replied:
I'd rather not make this change in the scope of this PR. I copied this code from `get_signing_address()`, and in this case I prefer consistency. We can change it after we merge these changes

reuvenpo commented on enigma-crypto/src/lib.rs:
```diff
@@ -58,3 +58,8 @@ pub trait Encryption<T, E, R, N>
+/// This object knows how to sign like a motherfucker
+pub trait EcdsaSign {
+    /// This object knows how to sign like a motherfucker
+    fn sign_hashed(&self, to_sign: &[u8; 32]) -> [u8; 65];
```
Instead of `[u8; 32]` you can accept web3::types::H256. I _think_ it's more appropriate semantically.

reuvenpo replied:
You should also call the method `sign_hash`

Cashmaney replied:
Why would we create a dependency on types that are defined by Web3? This is a generic ECDSA signing method

reuvenpo commented on enigma-principal/app/src/boot_network/principal_manager.rs:
```diff
@@ -64,6 +70,40 @@ pub struct PrincipalManager {
 }
 
+pub struct SgxEthereumSigner {
+    pub eid: sgx_enclave_id_t,
```
remove the pub, add a constructor (`new` method) and a getter (just called `eid`, not `get_eid`)

reuvenpo commented on enigma-principal/app/src/boot_network/principal_manager.rs:
```diff
@@ -64,6 +70,40 @@ pub struct PrincipalManager {
+}
+
+pub struct PrivateKeyEthereumSigner {
+    pub private_key: [u8; 32]
```
remove the pub, add a constructor (`new` method) and a getter (just called `private_key`, not `get_private_key`)

reuvenpo commented on enigma-principal/app/src/cli/app.rs:
```diff
@@ -1,32 +1,67 @@
+                        Sampler,
+                        PrincipalConfig,
+                        SgxEthereumSigner,
+                        PrivateKeyEthereumSigner},
```
Please stop formatting things this way...

reuvenpo replied:
instead do this:
```rust
principal_manager::{
    self,
    PrincipalManager,
    ReportManager,
    Sampler,
    PrincipalConfig,
    SgxEthereumSigner,
    PrivateKeyEthereumSigner,
},
```

Cashmaney replied:
I have no problem fixing it, but these are not things that are in the scope of this PR. When you're too aggressive with these formatting changes you end up with PRs that say "added local signing", and instead of being able to see the core code that was changed you end up with 25 files changed with random spacing stuff.

reuvenpo commented on enigma-principal/app/src/cli/app.rs:
```diff
@@ -1,32 +1,67 @@
+pub fn create_signer(eid: sgx_enclave_id_t, with_private_key: bool, private_key: Vec<u8>) -> Box<dyn EcdsaSign + Send + Sync> {
+    if with_private_key {
+        let mut pk_32 = [0u8; 32];
+        for (i, v) in private_key.iter().enumerate() { pk_32[i] = *v; }
```
instead of doing this, set the type of `private_key` in the fn signature to `&[u8]` and then use `pk_32.copy_from_slice(private_key)`
HINT: when calling this function, just add a `&` before the `private_key`. it will do the right conversion using `Deref`

reuvenpo replied:
Generally i don't love this function's signature but we can leave that for future refactoring.

Cashmaney replied:
> instead of doing this, set the type of `private_key` in the fn signature to `&[u8]` and then use `pk_32.copy_from_slice(private_key)`
> HINT: when calling this function, just add a `&` before the `private_key`. it will do the right conversion using `Deref`

Works for me also without changing the fn signature. Sounds okay?

reuvenpo commented on enigma-principal/app/src/cli/app.rs:
```diff
@@ -1,32 +1,67 @@
+    let ethereum_address = report_manager.get_ethereum_address()?;
     let mut path = storage_dir(ENCLAVE_DIR)?;
+    // let ethereum_signer = Arc::new(Mutex::new(Box::new(SgxEthereumSigner { eid }) as Box<dyn EcdsaSign + Send>));
+    // if principal_config.with_private_key {
```
remove this comment

reuvenpo commented on enigma-principal/app/src/cli/app.rs:
```diff
@@ -1,32 +1,67 @@
     if opt.info {
-        cli::options::print_info(&signing_address);
+        cli::options::print_info(&signing_address, &ethereum_address);
+        // cli::options::print_info(&ethereum_address);
```
remove comment

reuvenpo reviewed (commented):
Please merge the changes in #247 as well.
Please link to the PR in docker-environment that is associated with this PR.
After doing all your changes, run `cargo check --all-targets` and fix all errors warnings

reuvenpo commented on enigma-tools-u/Cargo.toml:
```diff
@@ -23,12 +24,16 @@ log = "0.4"
+ethereum-types = "0.6.0"
+rlp = "0.4.3"
+# ethereum-tx-sign = "2"
+# rlp2 = { package = "rlp", version = "0.4.2", optional = true }
```
remove comments

reuvenpo replied:
also remove secp256k1 and rlp as they are not used.

reuvenpo commented on enigma-tools-u/src/lib.rs:
```diff
@@ -24,13 +24,19 @@ extern crate log;
+// extern crate ethereum_tx_sign;
+//extern crate rlp;
+extern crate tiny_keccak;
+//extern crate secp256k1;
```
remove comments

reuvenpo commented on enigma-tools-u/src/web3_utils/contract_ext.rs:
```diff
@@ -0,0 +1,161 @@
+    let tx_hash = tx.to_hash(&chain_id);
+
+    let mut hashed = [0u8; 32];
+    for (i, v) in tx_hash.as_slice().iter().enumerate() { hashed[i] = *v; }
```
replace with `tx_hash.copy_from_slice(&hashed[..])`

reuvenpo commented on enigma-tools-u/src/web3_utils/contract_ext.rs:
```diff
@@ -0,0 +1,161 @@
+    println!("Hash: {:?}", tx_hash.to_hex());
+    println!("Hash: {:?}", sig[0..32].to_hex());
+    println!("Hash: {:?}", sig[32..64].to_hex());
+    println!("Hash: {:?}", v.to_hex());
```
remove printlns

reuvenpo commented on enigma-tools-u/src/web3_utils/contract_ext.rs:
```diff
@@ -0,0 +1,161 @@
+    // The actual calculation of V is [rec_id + chain * 2 + 35], but we expect v to already
+    // be (rec_id + 27) (see crypto::asymmetric)
+    let _v = chain_id * 2 + sig[64] as u64 + 35 - 27;
+    let v: Vec<u8> = [sig[64]].to_vec();
```
remove `v` and then rename `_v` to `v`

reuvenpo commented on enigma-tools-u/src/web3_utils/contract_ext.rs:
```diff
@@ -0,0 +1,161 @@
+    for (i, v) in tx_hash.as_slice().iter().enumerate() { hashed[i] = *v; }
+    let sig: [u8; 65] = signer.sign_hashed(&hashed);
+    // The actual calculation of V is [rec_id + chain * 2 + 35], but we expect v to already
+    // be (rec_id + 27) (see crypto::asymmetric)
```
great comment! :D

reuvenpo commented on enigma-tools-u/src/web3_utils/contract_ext.rs:
```diff
@@ -0,0 +1,161 @@
+use ethabi;
+use ethereum_types;
+use enigma_crypto::EcdsaSign;
+// use ethereum_tx_sign;
```
remove comment

reuvenpo commented on enigma-tools-u/src/web3_utils/contract_ext.rs:
```diff
@@ -0,0 +1,161 @@
+    },
+};
+use hex::{ToHex};
+/// I implement this struct to coalesce all the errors from the crates used in this file into one struct, that can
```
add one space after all imports

reuvenpo commented on enigma-tools-u/src/web3_utils/enigma_contract.rs:
```diff
@@ -88,35 +81,53 @@ pub trait ContractFuncs<G> {
+            "register",
+            (signing_address, report, signature),
+            opts,
+            self.chain_id, // TODO get this from configuration
```
remove todo

reuvenpo commented on enigma-tools-u/src/web3_utils/enigma_contract.rs:
```diff
@@ -88,35 +81,53 @@ pub trait ContractFuncs<G> {
+            "setWorkersParams",
+            (block_number, seed, sig.0),
+            opts,
+            self.chain_id, // TODO get this from configuration
```
remove todo

reuvenpo commented on enigma-tools-u/src/web3_utils/raw_transaction.rs:
```diff
@@ -0,0 +1,156 @@
+/// Description of a Transaction, pending or in the chain.
+#[derive(Debug, Default, Clone, PartialEq, Deserialize, Serialize)]
+pub struct RawTransaction {
+    /// Nonce
```
i have no idea if anything in this module is correct, but please remove all commented out code

reuvenpo commented on enigma-tools-u/src/web3_utils/raw_transaction.rs:
```diff
@@ -0,0 +1,156 @@
+    /// Signs and returns the RLP-encoded transaction - used to manually inject signature
+    pub fn raw_sign(&self, r: Vec<u8>, s: Vec<u8>, v: u64) -> Vec<u8> {
+        let mut r_n = r;
+        let mut s_n = s;
```
you can just put `mut` before the `r` and `s` in the function signature

Cashmaney replied:
I'd rather not start changing code that I copied :)

reuvenpo commented on enigma-tools-u/src/web3_utils/raw_transaction.rs:
```diff
@@ -0,0 +1,156 @@
+        }
+        while s_n[0] == 0 {
+            s_n.remove(0);
+        }
```
It will be much more efficient to .find() the first non-0 byte, and slice the vector from there.

Cashmaney replied:
Same as above

reuvenpo commented on enigma-tools-u/src/web3_utils/raw_transaction.rs:
```diff
@@ -0,0 +1,156 @@
+}
+
+fn keccak256_hash(bytes: &[u8]) -> Vec<u8> {
+    keccak256(bytes).into_iter().cloned().collect()
```
I'm pretty sure you can just write `keccak256(bytes).to_vec()`

Cashmaney replied:
Same as above:)

AvishaiW commented on enigma-principal/app/src/boot_network/keys_provider_http.rs:
```diff
@@ -176,15 +175,16 @@ impl PrincipalHttpServer {
     pub fn start(&self) {
         let mut io = IoHandler::default();
         let epoch_provider = Arc::clone(&self.epoch_provider);
+        let port = self.port;
```
what is the advantage of doing this?

reuvenpo replied:
The closure below is a `move` closure. If we don't dissociate `port` from `self` early, the closure will unnecessarily take ownership of all of `self`.

AvishaiW commented on enigma-principal/app/src/boot_network/principal_manager.rs:
```diff
@@ -64,6 +70,52 @@ pub struct PrincipalManager {
+}
+
+impl EcdsaSign for SgxEthereumSigner {
+    fn sign_hashed(&self, to_sign: &[u8; 32]) -> [u8; 65] {
```
what are the cases in which we are signing with the enclave and what are the cases which we are signing with the Ethereum Privkey? maybe it's worth adding documentation

reuvenpo replied:
As explained, this PR makes it so that whenever a key is not provided as part of the configuration, a key is instead generated in the enclave.
We should document this, but this isn't the right location. perhaps in the configuration struct used by KM? @Cashmaney ?

AvishaiW reviewed (approved):
Looking good!

reuvenpo commented:
running tests here:
https://drone.enigma.co/enigmampc/enigma-core/758

reuvenpo committed:
Added local signing of ethereum transactions made by the KM node.

This commit is huge, as it is a squash of several week's worth of work
done by myself (@reuvenpo) and @cashmaney. we do this because some of
the commits done by cashmaney were not signed/verified and the branch
configuration refused merging it all.

In this work we change the way ethereum functions are made against the
Enigma smart contract from the KM node, such that we no longer rely on
an unlocked account on the ethereum node we connect to, but rather sign
each request with a private key stored in the KM, which is associated
with a new ethereu address.

There are two possible ways of generating the keys and address for the KM node:

1. We generate a new keypair and address inside the SGX enclave, and expose the
   public key
2. We read the private key and address from the configuration.

One of these is necessarily chosen.
